src/context/WalletProvider/Ledger/components/Chains.tsx

```diff
@@ -0,0 +1,161 @@
+        )
+        const accountResults = await Promise.allSettled(accountPromises)
+
+        const balanceByChainId = accountResults.reduce<Record<ChainId, BN>>((acc, res, idx) => {
+          if (res.status === 'rejected') return acc
+          const { data: account } = res.value
+          if (!account) return acc
+
+          const accountId = accountIds[idx]
+          const { chainId } = fromAccountId(accountId)
+          const accountBalance = Object.values(account.accountBalances.byId).reduce(
+            (acc, byAssetId) => {
+              Object.values(byAssetId).forEach(balance => (acc = acc.plus(bnOrZero(balance))))
+              return acc
+            },
+            bnOrZero(0),
+          )
+
+          acc[chainId] = bnOrZero(acc[chainId]).plus(accountBalance)
+
+          const accountMetadata = accountMetadataByAccountId[accountId]
+          const payload = { [accountId]: accountMetadata }
+          dispatch(portfolio.actions.upsertAccountMetadata(payload))
+          dispatch(portfolio.actions.upsertPortfolio(account))
+          return acc
+        }, {})
+
+        Object.entries(balanceByChainId).forEach(([chainId, balance]) => {
+          if (balance.eq(0)) pull(availableChainIds, chainId)
+        })
+      } catch (e) {
+        console.error(e)
+      } finally {
+        setLoadingChains(prevLoading => ({ ...prevLoading, [chainId]: false }))
+      }
+    },
+    [availableChainIds, dispatch, walletState.adapters, walletState.wallet],
+  )
+
+  return (
+    <>
+      <ModalBody textAlign='left' pb={8}>
+        <Box mb={4}>
+          <Text
+            fontSize='lg'
+            fontWeight='bold'
+            translation={'walletProvider.ledger.chains.header'}
+            mb={2}
+          />
+          <Text color='text.subtle' translation={'walletProvider.ledger.chains.body'} />
+        </Box>
+        <Box mb={4}>
+          <Text
+            fontSize='md'
+            fontWeight='bold'
+            translation={'walletProvider.ledger.chains.availableChains'}
+            mb={2}
+          />
+          <Box>
+            {availableAssets
+              .sort(a => (walletChainIds.includes(a.chainId) ? 1 : -1))
+              .map(asset => (
+                <Flex alignItems='center' justifyContent='space-between' mb={4} key={asset.assetId}>
+                  <Flex alignItems='center'>
+                    <AssetIcon assetId={asset.assetId} mr={2} />
+                    <CText>{asset.name}</CText>
+                  </Flex>
+                  {walletChainIds.includes(asset.chainId) ? (
+                    <CText>Added</CText>
+                  ) : loadingChains[asset.chainId] ? (
+                    <Spinner />
+                  ) : (
+                    <Button colorScheme='blue' onClick={() => handleConnectClick(asset.chainId)}>
+                      Connect
+                    </Button>
+                  )}
+                </Flex>
```

Comment:
not sure how fussy we want to be on performance here, but if its worth it you can break this out into a mini component

Reply:
how about this 5039c7754fa1e6741de3e0372f952cb7d7a0b54